_posts/2019-01-14-Infrastructure-as-code-with-terraform-and-aws-serverless.md

```diff
@@ -0,0 +1,497 @@
+The output from the hash function determines the partition (physical storage internal to DynamoDB) in which the item will be stored. 
+All items with the same partition key value are stored together, in sorted order by sort key value.
+-- from AWS docs: <a target="_blank" href="https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.CoreComponents.html">https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.CoreComponents.html</a>
+
+From the attribute definitions in dynamo.tf it is clear that Author (S) is a string and Date (N) should be a number.
+
+# IAM
+<div style="text-align: center;">
+  <img src="/img/2019-01-14-Infrastructure-as-code-with-terraform-and-aws-serverless/icon-IAM.png" width="15%" height="15%">
+</div>
+
+Before specifying the Lambda Functions we create all the permissions for our functions.
+Basically, this is making sure that our functions have permissions to access other resources (like DynamoDB).
+Without going to deep into it the aws permission model works as follows:
+* you give a resource a role
+* to this role you add permissions
+* these permissions give it access to other resources:
+    * either permission for triggering an other resource (eg. Lambda Function forwards logs to CloudWatch)
+    * or permission for being triggered by an other resource (eg. Lambda Function may be triggered by ApiGateway)
+
+```hcl-terraform
+# ROLES
+# IAM role which dictates what other AWS services the Lambda function
+# may access.
+resource "aws_iam_role" "lambda-iam-role" {
+  name = "codingtips_lambda_role"
+
+  assume_role_policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Action": "sts:AssumeRole",
+      "Principal": {
+        "Service": "lambda.amazonaws.com"
+      },
+      "Effect": "Allow",
+      "Sid": ""
+    }
+  ]
+}
+EOF
+}
+
+# POLICIES
+resource "aws_iam_role_policy" "dynamodb-lambda-policy"{
+  name = "dynamodb_lambda_policy"
+  role = "${aws_iam_role.lambda-iam-role.id}"
+  policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "dynamodb:*"
+      ],
+      "Resource": "${aws_dynamodb_table.codingtips-dynamodb-table.arn}"
+    }
+  ]
+}
+EOF
+}
+
+resource "aws_iam_role_policy" "cloudwatch-lambda-policy"{
+  name = "cloudwatch-lambda-policy"
+  role = "${aws_iam_role.lambda-iam-role.id}"
+  policy = "${data.aws_iam_policy_document.api_gateway_logs_policy_document.json}"
+}
+
+data "aws_iam_policy_document" "api_gateway_logs_policy_document" {
+  statement {
+    actions = [
+      "logs:CreateLogStream",
+      "logs:CreateLogGroup",
+      "logs:PutLogEvents"
+    ],
+    resources = [
+      "arn:aws:logs:*:*:*"
+    ]
+  }
+}
+
+```
+
+In the example above the first resource that is defined is an `aws_iam_role`.
+This is the role that we will later give to our Lambda Functions.
+
+Then, we create `aws_iam_role_policy` resources to attach the role.
```

Comment:
```suggestion
We then create the `aws_iam_role_policy` policy to attach to the role.
```
Not sure if I understood this correct?

Reply:
I will change the syntax. I am referring to the "aws_iam_role_policy" "dynamodb-lambda-policy"  in which we state role = "${aws_iam_role.lambda-iam-role.id}" to couple both resources.